src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs

```diff
@@ -84,20 +84,33 @@ class NoUnexpectedThunks a where
   --
   -- See also discussion of caveats listed for 'NF.isNormalForm'.
   noUnexpectedThunks :: HasCallStack => [String] -> a -> IO ThunkInfo
-  noUnexpectedThunks ctxt x = do
-      c   <- getBoxedClosureData (asBox x)
-      hnf <- NF.isHeadNormalForm c
-      if hnf
-        then whnfNoUnexpectedThunks ctxt' x
-        else return $ UnexpectedThunk UnexpectedThunkInfo {
-                          unexpectedThunkContext   = ctxt'
-                        , unexpectedThunkCallStack = callStack
-                        , unexpectedThunkClosure   = Just (renderClosure c)
-                        }
+  noUnexpectedThunks ctxt x
+    | length ctxt' <= maxContextLength = do
+        c   <- getBoxedClosureData (asBox x)
+        hnf <- NF.isHeadNormalForm c
+        if hnf
+          then whnfNoUnexpectedThunks ctxt' x
+          else return $ UnexpectedThunk UnexpectedThunkInfo {
+                            unexpectedThunkContext   = ctxt'
+                          , unexpectedThunkCallStack = callStack
+                          , unexpectedThunkClosure   = Just (renderClosure c)
+                          }
+    | otherwise = do
+        return $ UnexpectedThunk UnexpectedThunkInfo {
+            unexpectedThunkContext   = "<<loop?>>" : ctxt'
+          , unexpectedThunkCallStack = callStack
+          , unexpectedThunkClosure   = Nothing
+          }
     where
       ctxt' :: [String]
       ctxt' = showTypeOf (Proxy @a) : ctxt
 
+      -- If we exceed the maximum context length, it is likely that there is a
+      -- loop in a 'NoUnexpectedContext' instance somewhere; it is unlikely that
+      -- a type is so complicated it exceeds the maximum context length.
+      maxContextLength :: Int
+      maxContextLength = 1000
```

Comment:
Should the `maxContextLength` instead be specified by the caller?

Reply:
I guess that's not unreasonable.

Reply:
I think this is okay for now but just a thought.